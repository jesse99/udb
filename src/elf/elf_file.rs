//! Data within a core file or exe.
use super::{
    ElfHeader, LoadSegment, MemoryMappedFile, NoteType, PrStatus, ProgramHeader, Reader,
    SectionIndex, SegmentType, Stream,
};
use crate::debug::{LineInfo, SymbolTable, SymbolTableEntry};
use crate::elf::{
    Bytes, ChildSignal, CoreNoteType, FaultSignal, KillSignal, Note, Offset, PosixSignal,
    RelativeAddr, Relocation, SectionHeader, SectionType, SigInfo, SignalDetails, StringIndex,
    VirtualAddr,
};
use crate::utils::{self, warn};
use memmap2::Mmap;
use std::cell::OnceCell;
use std::error::Error;
use std::fs::File;

pub struct ElfFile {
    pub header: ElfHeader,
    pub path: std::path::PathBuf,
    pub reader: &'static Reader,
    pub loads: Vec<LoadSegment>,
    pub notes: Vec<Note>,
    pub sections: Vec<SectionHeader>, // not used for core files

    memory_mapped: OnceCell<Option<Vec<MemoryMappedFile>>>,
    lines: OnceCell<Option<LineInfo>>,
}

impl ElfFile {
    pub fn new(path: std::path::PathBuf) -> Result<Self, Box<dyn Error>> {
        let file = File::open(path.clone())?;

        // This is unfafe because it has undefined behavior if the underlying file is
        // modified while the memory map is in use.
        let bytes = unsafe { Mmap::map(&file) }?;

        // Readers will always persist for the entire life of udb so it's OK to leak them.
        // And because we leak them we can use stuff like StringView containing a static
        // reference to a Reader allowing us to not allocate memory for strings until we
        // actually need it.
        let reader = Box::leak(Box::new(Reader::new(bytes)?));
        let header = ElfHeader::new(reader)?;
        let loads = ElfFile::load_loads(reader, &header);
        let notes = ElfFile::load_notes(reader, &header);
        let sections = ElfFile::load_sections(reader, &header);
        ElfFile::load_others(reader, &header);
        Ok(ElfFile {
            path,
            reader,
            header,
            loads,
            notes,
            sections,
            memory_mapped: OnceCell::new(),
            lines: OnceCell::new(),
        })
    }

    pub fn find_load_segment(&self, vaddr: VirtualAddr) -> Option<&LoadSegment> {
        self.loads.iter().find(|s| s.vbytes.contains(vaddr))
    }

    pub fn offset_to_vaddr(&self, offset: Offset) -> Option<(&LoadSegment, VirtualAddr)> {
        self.loads
            .iter()
            .find(|s| s.obytes.contains(offset))
            .map(|s| (s, s.vbytes.start + (offset - s.obytes.start)))
    }

    pub fn vaddr_to_raddr(&self, addr: VirtualAddr) -> Option<RelativeAddr> {
        let files = self.get_memory_mapped_files();
        files.as_ref().and_then(|maps| {
            maps.iter()
                .find(|s| s.vbytes.contains(addr))
                .map(|s| RelativeAddr(addr.0 - s.vbytes.start.0))
        })
    }

    /// Returns a string from the section string table. Note that index can point into
    /// the middle of a string.
    pub fn find_default_string(&self, index: StringIndex) -> Option<String> {
        let section = SectionIndex(self.header.string_table_index as u32);
        self.find_string(section, index)
    }

    /// Returns a string from an arbitrary string table. Note that index can point into
    /// the middle of a string.
    pub fn find_string(&self, section: SectionIndex, index: StringIndex) -> Option<String> {
        let h = self.find_section(section)?;
        // TODO really should return an error if indexing past h.offset + h.size
        let offset = h.obytes.start + index.0 as i64;
        match Stream::new(self.reader, offset).read_string() {
            Ok(s) => Some(s),
            Err(err) => {
                utils::warn(&format!("failed to read section string {index:?}: {err}"));
                None
            }
        }
    }

    pub fn find_strings(&self, section: &SectionHeader, max: usize) -> Vec<String> {
        let mut result = Vec::new();
        if section.stype == SectionType::StringTable {
            let mut stream = Stream::new(self.reader, section.obytes.start);
            while stream.offset < section.obytes.end() {
                if let Ok(s) = stream.read_string() {
                    result.push(s);
                    if result.len() == max {
                        break;
                    }
                }
            }
        }
        result
    }

    pub fn find_section_name(&self, section: SectionIndex) -> Option<String> {
        let h = self.find_section(section)?;
        self.find_default_string(h.name)
    }

    pub fn get_lines(&self) -> &Option<LineInfo> {
        self.lines.get_or_init(|| {
            for (i, section) in self.sections.iter().enumerate() {
                if section.stype == SectionType::ProgBits {
                    let index = SectionIndex(i as u32);
                    if let Some(name) = self.find_section_name(index)
                        && name == ".debug_line"
                    {
                        let max_offset = section.obytes.end();
                        return Some(LineInfo::new(
                            &mut Stream::new(self.reader, section.obytes.start),
                            max_offset,
                        ));
                    }
                }
            }
            None
        })
    }

    pub fn find_symbols(&self) -> Option<SymbolTable> {
        self.do_find_symbols(SectionType::SymbolTable)
    }

    pub fn find_dynamic_symbols(&self) -> Option<SymbolTable> {
        self.do_find_symbols(SectionType::DynamicSymbolTable)
    }

    pub fn find_segments(reader: &'static Reader, header: &ElfHeader) -> Vec<ProgramHeader> {
        let mut segments = Vec::new();
        let mut offset = Offset(header.ph_offset);

        for _ in 0..header.num_ph_entries {
            match ProgramHeader::new(reader, offset) {
                Ok(ph) => segments.push(ph),
                Err(err) => {
                    utils::warn(&format!(
                        "failed to read program header at {offset:?}: {err}"
                    ));
                }
            }
            offset = offset + header.ph_entry_size as i64;
        }
        segments
    }

    pub fn get_sections(&self) -> &Vec<SectionHeader> {
        &self.sections
    }

    pub fn get_memory_mapped_files(&self) -> &Option<Vec<MemoryMappedFile>> {
        fn get_files(s: &mut Stream) -> Result<Vec<MemoryMappedFile>, Box<dyn Error>> {
            // For some reason files get mapped in multiple times, e.g.
            //    7f45e7402000 7f45e7559000   1404928 /usr/lib64/libxpath.so
            //    7f45e7559000 7f45e7758000   2093056 /usr/lib64/libxpath.so
            //    7f45e7758000 7f45e77cd000    479232 /usr/lib64/libxpath.so
            //    7f45e77cd000 7f45e7a37000   2531328 /usr/lib64/libxpath.so
            // This is annoying and not useful so we we'll merge them together.
            // Note that the end of one line usually matches the start of the next.
            let count = s.read_ulong()?;
            let _page_size = s.read_ulong()?;

            let mut elements = Vec::new();
            for _ in 0..count {
                let start = s.read_ulong()?;
                let end = s.read_ulong()?;
                let offset = s.read_ulong()?;
                elements.push((start, end, offset));
            }

            let mut files: Vec<MemoryMappedFile> = Vec::new();
            for (start, end, _offset) in elements {
                if let Ok(file_name) = s.read_string() {
                    if let Some(old) = files.last_mut()
                        && start == old.vbytes.end().0
                        && file_name == old.file_name
                    {
                        old.vbytes.size = (end - old.vbytes.start.0) as usize;
                    } else {
                        files.push(MemoryMappedFile {
                            vbytes: Bytes::<VirtualAddr>::from_raw(start, (end - start) as usize),
                            // offset: offset * page_size,
                            file_name,
                        });
                    }
                } else {
                    utils::warn(&format!(
                        "Failed to read MemoryMappedFile at offset {:?}",
                        s.offset
                    ));
                }
            }

            Ok(files)
        }

        self.memory_mapped.get_or_init(|| {
            if let Some(note) = self.find_core_note(CoreNoteType::File) {
                let mut s = Stream::new(self.reader, note.contents.start);
                match get_files(&mut s) {
                    Ok(files) => Some(files),
                    Err(e) => {
                        utils::warn(&format!("Error reading memory mapped files: {}", e));
                        None
                    }
                }
            } else {
                None
            }
        })
    }

    pub fn find_core_note(&self, ntype: CoreNoteType) -> Option<&Note> {
        for note in self.notes.iter() {
            if let NoteType::Core(t) = &note.ntype
                && *t == ntype
            {
                return Some(note);
            }
        }
        None
    }

    pub fn find_prstatus(&self) -> Option<PrStatus> {
        fn get_prstatus(s: &mut Stream) -> Result<PrStatus, Box<dyn Error>> {
            // See elf_prstatus in https://docs.huihoo.com/doxygen/linux/kernel/3.7/uapi_2linux_2elfcore_8h_source.html
            let signal_num = s.read_int()?;
            let signal_code = s.read_int()?;
            let _errno = s.read_int()?;
            let _current_signal = s.read_half()?; // This is the current signal, not the one that caused the core dump.
            let _padding = s.read_half()?;
            let _pending_signals = s.read_xword()?;
            let _held_signals = s.read_xword()?;
            let pid = s.read_int()?;
            let _pppid = s.read_int()?;
            let _pgrp = s.read_int()?;
            let _prsid = s.read_int()?;

            let _utime_s = s.read_xword()?; // time spent in user code
            let _utime_u = s.read_xword()?;

            let _stime_s = s.read_xword()?; // time spent in system code
            let _stime_u = s.read_xword()?;

            let _cutime_s = s.read_xword()?;
            let _cutime_u = s.read_xword()?;

            let _cstime_s = s.read_xword()?;
            let _cstime_u = s.read_xword()?;

            // TODO good only for x86 and arm
            let mut registers = Vec::new();
            for _ in 1..27 {
                let r = s.read_xword()?;
                registers.push(r);
            }
            // TODO may need to use pr_exec_fdpic_loadmap

            Ok(PrStatus {
                signal_num,
                signal_code,
                // errno,
                pid,
                registers,
            })
        }

        if let Some(note) = self.find_core_note(CoreNoteType::PrStatus) {
            let mut s = Stream::new(self.reader, note.contents.start);
            match get_prstatus(&mut s) {
                Ok(status) => Some(status),
                Err(e) => {
                    utils::warn(&format!("Error reading prstatus: {}", e));
                    None
                }
            }
        } else {
            None
        }
    }

    pub fn find_signal_info(&self) -> Option<SigInfo> {
        fn get_signal_info(s: &mut Stream) -> Result<SigInfo, Box<dyn Error>> {
            const SI_MASK: u32 = 0xffff0000;
            const SI_KILL: u32 = 0 << 16;
            const SI_TIMER: u32 = 1 << 16;
            const SI_POLL: u32 = 2 << 16;
            const SI_FAULT: u32 = 3 << 16;
            const SI_CHLD: u32 = 4 << 16;
            const SI_MESGQ: u32 = 6 << 16;
            const SI_SYS: u32 = 7 << 16;

            // See https://elixir.bootlin.com/linux/v4.9/source/arch/ia64/include/uapi/asm/siginfo.h#L83
            // and https://elixir.bootlin.com/linux/v4.9/source/arch/arm64/kernel/signal32.c#L144
            let _si_signo = s.read_xword()? as i32; // TODO I think these fields are all emitted as 64 bits?
            let _si_errno = s.read_xword()? as i32; // TODO need to test this better
            let _si_code = s.read_xword()? as i32; // TODO this seems completely wrong: we're getting kill for a seg fault

            let details = match (_si_code as u32) & SI_MASK {
                SI_KILL => {
                    let sender_pid = s.read_xword()? as i32;
                    let sender_uid = s.read_xword()? as i32;
                    SignalDetails::Kill(KillSignal {
                        sender_pid,
                        sender_uid,
                    })
                }
                SI_TIMER => SignalDetails::Timer, // TODO bit more we could include here
                SI_POLL => SignalDetails::Poll,   // TODO bit more we could include here
                SI_FAULT => {
                    let fault_addr = s.read_addr()?;
                    SignalDetails::Fault(FaultSignal { fault_addr })
                }
                SI_CHLD => {
                    let child_pid = s.read_xword()? as i32;
                    let child_uid = s.read_xword()? as i32;
                    let exit_code = s.read_xword()? as i32;
                    SignalDetails::Child(ChildSignal {
                        child_pid,
                        child_uid,
                        exit_code,
                    })
                }
                SI_MESGQ => SignalDetails::MesgQ, // TODO more we can add here
                SI_SYS => SignalDetails::Sys,     // TODO more we can add here
                _ => {
                    let sender_pid = s.read_xword()? as i32;
                    let sender_uid = s.read_xword()? as i32;
                    SignalDetails::Posix(PosixSignal {
                        sender_pid,
                        sender_uid,
                    })
                }
            };

            Ok(SigInfo {
                // signal_num: si_signo,
                // errno: si_errno,
                // signal_code: si_code,
                details,
            })
        }

        if let Some(note) = self.find_core_note(CoreNoteType::SigInfo) {
            let mut s = Stream::new(self.reader, note.contents.start);
            match get_signal_info(&mut s) {
                Ok(status) => Some(status),
                Err(e) => {
                    utils::warn(&format!("Error reading signal info: {}", e));
                    None
                }
            }
        } else {
            None
        }
    }

    pub fn find_relocations(&self, result: &mut Vec<Relocation>) {
        fn load_with(reader: &'static Reader, offset: Offset, dynamic: bool) -> Option<Relocation> {
            match Relocation::with_addend(reader, offset, dynamic) {
                Ok(r) => Some(r),
                Err(err) => {
                    utils::warn(&format!("couldn't read relocation at {offset:?}: {err}"));
                    None
                }
            }
        }

        fn load_without(
            reader: &'static Reader,
            offset: Offset,
            dynamic: bool,
        ) -> Option<Relocation> {
            match Relocation::with_no_addend(reader, offset, dynamic) {
                Ok(r) => Some(r),
                Err(err) => {
                    utils::warn(&format!("couldn't read relocation at {offset:?}: {err}"));
                    None
                }
            }
        }

        fn load_relocations_with(
            reader: &'static Reader,
            section: &SectionHeader,
            result: &mut Vec<Relocation>,
        ) {
            let mut offset = section.obytes.start;
            while offset + section.entry_size as i64 <= section.obytes.end() {
                let dynamic = section.info == 0; // TODO better to look at section name?
                if let Some(r) = load_with(reader, offset, dynamic) {
                    result.push(r)
                }
                offset = offset + section.entry_size as i64;
            }
        }

        fn load_relocations_without(
            reader: &'static Reader,
            section: &SectionHeader,
            result: &mut Vec<Relocation>,
        ) {
            let mut offset = section.obytes.start;
            while offset + section.entry_size as i64 <= section.obytes.end() {
                let dynamic = section.info == 0; // TODO better to look at section name?
                if let Some(r) = load_without(reader, offset, dynamic) {
                    result.push(r)
                }
                offset = offset + section.entry_size as i64;
            }
        }

        if !self.header.is_x66_64() {
            utils::warn("relocations are only supported for x86 64-bit");
            return;
        }
        for section in self.sections.iter() {
            match section.stype {
                SectionType::RelocationsWith => load_relocations_with(self.reader, section, result),
                SectionType::RelocationsWithout => {
                    load_relocations_without(self.reader, section, result)
                }
                _ => (),
            }
        }
    }
}

impl ElfFile {
    fn find_section(&self, section: SectionIndex) -> Option<&SectionHeader> {
        let section_index = section.0 as usize;
        let section = self.sections.get(section_index);
        if section.is_none() {
            utils::warn(&format!("bad section index: {section_index}"));
        }
        section
    }

    fn do_find_symbols(&self, stype: SectionType) -> Option<SymbolTable> {
        for section in self.sections.iter() {
            if section.stype == stype {
                // TODO warn if there is more than one of these
                let mut offset = section.obytes.start;
                let mut entries = Vec::new();
                while offset < section.obytes.end() {
                    match SymbolTableEntry::new(self.reader, offset) {
                        Ok(s) => entries.push(s),
                        Err(err) => warn(&format!(
                            "failed to read symbols at offset {offset:?}: {err}"
                        )),
                    }
                    offset = offset + section.entry_size as i64;
                }
                let table = SymbolTable {
                    section: section.clone(),
                    dynamic: stype == SectionType::DynamicSymbolTable,
                    entries,
                };
                return Some(table);
            }
        }
        None
    }

    fn load_loads(reader: &'static Reader, header: &ElfHeader) -> Vec<LoadSegment> {
        let mut loads = Vec::new();
        let mut offset = Offset(header.ph_offset);

        // Even a large core file has a small number of program headers, so it's OK to
        // re-iterate over them.
        for _ in 0..header.num_ph_entries {
            match ProgramHeader::new(reader, offset) {
                Ok(ph) => {
                    if ph.stype == SegmentType::Load {
                        let obytes = Bytes::<Offset>::from_raw(ph.offset, ph.mem_size as usize);
                        let vbytes = Bytes::<VirtualAddr>::from_raw(ph.vaddr, ph.mem_size as usize);
                        loads.push(LoadSegment {
                            obytes,
                            vbytes,
                            // paddr: ph.paddr,
                            flags: ph.flags,
                        });
                    }
                }
                Err(err) => utils::warn(&format!(
                    "failed to read program header at {offset:?}: {err}"
                )),
            }
            offset = offset + header.ph_entry_size as i64;
        }
        loads
    }

    fn load_notes(reader: &'static Reader, header: &ElfHeader) -> Vec<Note> {
        fn load_note(s: &mut Stream) -> Option<Note> {
            if let Ok((name, ntype, contents)) = super::read_note(s) {
                Some(Note {
                    name: name.clone(),
                    ntype: NoteType::new(&name, ntype),
                    contents,
                })
            } else {
                utils::warn(&format!("Failed to read note at offset {:?}", s.offset));
                None
            }
        }

        let mut notes = Vec::new();
        let mut offset = Offset(header.ph_offset);

        for _ in 0..header.num_ph_entries {
            match ProgramHeader::new(reader, offset) {
                Ok(ph) => {
                    // Note that core files can sometimes be damaged (typically because they are
                    // truncated). Not all notes are essential so we'll try to continue even if
                    // a note cannot be read.
                    if ph.stype == SegmentType::Note {
                        let start = Offset(ph.offset);
                        let end = start + ph.file_size as i64;
                        let mut s = Stream::new(reader, start);
                        while s.offset < end {
                            if let Some(note) = load_note(&mut s) {
                                // TODO may want to warn if get a second note of the same type
                                notes.push(note);
                            }
                        }
                    }
                }
                Err(err) => utils::warn(&format!(
                    "failed to read program header at {offset:?}: {err}"
                )),
            }
            offset = offset + header.ph_entry_size as i64;
        }
        notes
    }

    // This is just here so we can report unknown segments.
    fn load_others(reader: &'static Reader, header: &ElfHeader) {
        let mut offset = Offset(header.ph_offset);

        for _ in 0..header.num_ph_entries {
            match ProgramHeader::new(reader, offset) {
                Ok(ph) => match ph.stype {
                    SegmentType::Dynamic => (), // TODO may need to use this one
                    SegmentType::Interpreter => (),
                    SegmentType::Note => (),
                    SegmentType::Null => (),
                    SegmentType::Load => (),
                    SegmentType::Phdr => (),
                    _ => utils::warn(&format!("Ignoring segment type {:?}", ph.stype)),
                },
                Err(err) => utils::warn(&format!(
                    "failed to read program header at {offset:?}: {err}"
                )),
            }
            offset = offset + header.ph_entry_size as i64;
        }
    }

    fn load_sections(reader: &'static Reader, header: &ElfHeader) -> Vec<SectionHeader> {
        let mut sections = Vec::new();
        let mut offset = Offset(header.section_offset);

        for _ in 0..header.num_section_entries {
            match SectionHeader::new(reader, offset) {
                Ok(h) => sections.push(h),

                Err(err) => utils::warn(&format!(
                    "failed to read section header at {offset:?}: {err}"
                )),
            }
            offset = offset + header.section_entry_size as i64;
        }
        sections
    }
}

// TODO may want to add an option to Reader so that it (randomly?) fails to see how we
// handle corrupt files, or maybe better to simulate truncation
#[cfg(test)]
mod tests {
    use super::*;

    // TODO duplicate these for the release core? or is there something we can do to
    // parameterize it with insta? or maybe just have the tests do both cores?
    #[test]
    fn debug_header() {
        let path = std::path::PathBuf::from("cores/shopping-debug/app-debug.core");
        let core = ElfFile::new(path).unwrap();
        let s = format!("{} on {}", core.header.machine(), core.header.abi());
        insta::assert_snapshot!(s);
    }

    #[test]
    fn debug_signal() {
        let path = std::path::PathBuf::from("cores/shopping-debug/app-debug.core");
        let core = ElfFile::new(path).unwrap();
        let s = match core.find_prstatus() {
            Some(status) => status.signal(),
            None => "no pr status",
        };
        insta::assert_snapshot!(s);
    }

    #[test]
    fn debug_memory_mapped_files() {
        let path = std::path::PathBuf::from("cores/shopping-debug/app-debug.core");
        let core = ElfFile::new(path).unwrap();
        let s = match core.get_memory_mapped_files() {
            // start_addr and end_addr will change with each build
            // size will change if we tweak the code
            // so we won't test any of that
            Some(files) => files.iter().map(|f| format!("{}\n", f.file_name)).collect(),
            None => "no files".to_string(),
        };
        insta::assert_snapshot!(s);
    }
}
