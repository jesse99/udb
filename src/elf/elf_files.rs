use crate::{
    debug::LineInfo,
    elf::{ElfFile, LoadSegment, PrStatus, Relocation, VirtualAddr},
};
use std::error::Error;

pub struct ElfFiles {
    pub core: Option<ElfFile>,
    pub exe: Option<ElfFile>,
}

impl ElfFiles {
    pub fn new(paths: Vec<std::path::PathBuf>) -> Result<Self, Box<dyn Error>> {
        let files = paths
            .into_iter()
            .map(|p| ElfFile::new(p))
            .collect::<Result<Vec<_>, _>>()?;
        let mut core = None;
        let mut exe = None;
        for file in files {
            if file.header.etype == 4 {
                if core.is_none() {
                    core = Some(file);
                } else {
                    return Err("can't have multiple core files".into());
                }
            } else if exe.is_none() {
                exe = Some(file);
            } else {
                return Err("can't have multiple exe files".into());
            }
        }
        Ok(ElfFiles { core, exe })
    }

    pub fn find_load_segment(&self, vaddr: VirtualAddr) -> Option<&LoadSegment> {
        match &self.core {
            Some(c) => c.find_load_segment(vaddr),
            None => None,
        }
    }

    pub fn find_prstatus(&self) -> Option<PrStatus> {
        match &self.core {
            Some(c) => c.find_prstatus(),
            None => None,
        }
    }

    // pub fn find_vaddr(&self, offset: u64) -> Option<(&LoadSegment, u64)> {
    //     match &self.core {
    //         Some(c) => c.find_vaddr(offset),
    //         None => None,
    //     }
    // }

    /// Returns file name, line number, and column for the given address.
    pub fn find_line(&self, addr: VirtualAddr) -> Result<(String, u32, u16), Box<dyn Error>> {
        match (&self.core, &self.exe) {
            (Some(core), Some(exe)) => {
                match core.vaddr_to_raddr(addr) {
                    Some(addr) => {
                        match exe.get_lines() {
                            // TODO need to cache lines
                            Some(lines) => match lines.lines.get(&addr) {
                                Some(value) => {
                                    let file = lines.files.get(value.file);
                                    Ok((file.clone(), value.line, value.column))
                                }
                                None => Ok(("?".to_string(), 0, 0)),
                            },
                            None => Err("Couldn't find .debug_line section".into()),
                        }
                    }
                    None => Err("couldn't find a load segment matching the addr".into()),
                }
            }
            (None, Some(_)) => Err("need an core file to find file and line".into()),
            (Some(_), None) => Err("need an exe file to find file and line".into()), // TODO addr2line doesn't need a core file
            (None, None) => Err("need core and exe files to find file and line".into()),
        }
    }

    pub fn find_relocations(&self) -> Vec<Relocation> {
        let mut result = Vec::new();
        if let Some(file) = &self.core {
            file.find_relocations(&mut result);
        }
        if let Some(file) = &self.exe {
            file.find_relocations(&mut result);
        }
        result
    }
}
