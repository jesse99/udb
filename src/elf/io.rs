use crate::repl::HexdumpOffsets;
use crate::utils;
use crate::utils::Styling;
use crate::utils::print_styled;
use memmap2::Mmap;
use std::error::Error;

pub struct Reader {
    pub little_endian: bool,
    pub sixty_four_bit: bool,
    bytes: Mmap,
}

impl Reader {
    /// Note that these functions all return a Result because core files are sometimes
    /// corrupted and we want to continue to work as well as we can when that happens.
    pub fn new(bytes: Mmap) -> Result<Self, Box<dyn Error>> {
        // see https://en.wikipedia.org/wiki/Executable_and_Linkable_Format
        utils::require(bytes.len() > 64, "core file is much too small")?;
        let magic = bytes.get(0..4).unwrap();
        utils::require(
            magic[0] == 0x7f && magic[1] == 0x45 && magic[2] == 0x4c || magic[3] == 0x46,
            "not a core file (bad magic)",
        )?;

        let ei_class = *bytes.get(0x04).unwrap();
        let ei_data = *bytes.get(0x05).unwrap();
        let ei_version = *bytes.get(0x06).unwrap();
        let e_type = *bytes.get(0x10).unwrap();
        utils::require(ei_version == 1, &format!("bad elf version: {ei_version}"))?;
        utils::require(
            e_type == 0x02 || e_type == 0x03 || e_type == 0x04,
            "bad elf type: not a core, exe, or shared lib",
        )?;

        Ok(Reader {
            bytes,
            sixty_four_bit: ei_class == 2,
            little_endian: ei_data == 1,
        })
    }

    pub fn len(&self) -> usize {
        self.bytes.len()
    }

    pub fn slice(&self, offset: usize, size: usize) -> Result<&[u8], Box<dyn Error>> {
        if offset + size > self.bytes.len() {
            return Err("slice out of bounds".into());
        }
        Ok(&self.bytes[offset..offset + size])
    }

    pub fn read_byte(&self, offset: usize) -> Result<u8, Box<dyn Error>> {
        self.bytes
            .get(offset)
            .ok_or("couldn't read byte at offset".into())
            .copied()
    }

    pub fn read_half(&self, offset: usize) -> Result<u16, Box<dyn Error>> {
        let slice = &self.bytes[offset..offset + 2];
        if self.little_endian {
            Ok(u16::from_le_bytes(slice.try_into()?))
        } else {
            Ok(u16::from_be_bytes(slice.try_into()?))
        }
    }

    pub fn read_word(&self, offset: usize) -> Result<u32, Box<dyn Error>> {
        let slice = &self.bytes[offset..offset + 4];
        if self.little_endian {
            Ok(u32::from_le_bytes(slice.try_into()?))
        } else {
            Ok(u32::from_be_bytes(slice.try_into()?))
        }
    }

    pub fn read_xword(&self, offset: usize) -> Result<u64, Box<dyn Error>> {
        let slice = &self.bytes[offset..offset + 8];
        if self.little_endian {
            Ok(u64::from_le_bytes(slice.try_into()?))
        } else {
            Ok(u64::from_be_bytes(slice.try_into()?))
        }
    }

    /// Read either a u32 or u64 word depending on whether the core file is 64-bit.
    /// But, for sanity, always return the result as 64 bits.
    pub fn read_addr(&self, offset: usize) -> Result<u64, Box<dyn Error>> {
        if self.sixty_four_bit {
            self.read_xword(offset)
        } else {
            Ok(self.read_word(offset)? as u64)
        }
    }

    // TODO should address and offset be new types?
    fn read_offset(&self, offset: usize) -> Result<u64, Box<dyn Error>> {
        if self.sixty_four_bit {
            self.read_xword(offset)
        } else {
            Ok(self.read_word(offset)? as u64)
        }
    }

    pub fn hex_dump(&self, addr: u64, offset: usize, size: usize, offsets: HexdumpOffsets) {
        let mut i = offset;
        loop {
            match offsets {
                HexdumpOffsets::None => (),
                HexdumpOffsets::Addr => {
                    print_styled!("{:012x}: ", hex_offset, addr + (i - offset) as u64);
                }
                HexdumpOffsets::Zero => {
                    print_styled!("{:04x}: ", hex_offset, i - offset);
                }
            }

            for j in 0..8 {
                if i + j >= offset + size || i + j >= self.len() {
                    break;
                }
                print_styled!("{:02x} ", hex_hex, self.read_byte(i + j).unwrap());
            }
            print!(" ");
            for j in 0..8 {
                if i + j >= offset + size || i + j >= self.len() {
                    break;
                }
                print_styled!("{:02x} ", hex_hex, self.read_byte(i + j).unwrap());
            }
            print!("   ");
            for j in 0..16 {
                if i + j >= offset + size || i + j >= self.len() {
                    break;
                }
                let ch = self.read_byte(i + j).unwrap() as char;
                if ch.is_ascii_graphic() {
                    print_styled!("{ch}", hex_ascii);
                } else {
                    print_styled!(".", hex_ascii);
                }
            }
            println!();
            i += 16;
            if i >= offset + size || i >= self.len() {
                break;
            }
        }
    }
}

pub struct Stream<'a> {
    pub reader: &'a Reader,
    pub offset: usize,
}

impl<'a> Stream<'a> {
    pub fn new(reader: &'a Reader, offset: usize) -> Self {
        Stream { reader, offset }
    }

    pub fn read_byte(&mut self) -> Result<u8, Box<dyn Error>> {
        let byte = self.reader.read_byte(self.offset)?;
        self.offset += 1;
        Ok(byte)
    }

    pub fn read_half(&mut self) -> Result<u16, Box<dyn Error>> {
        let half = self.reader.read_half(self.offset)?;
        self.offset += 2;
        Ok(half)
    }

    pub fn read_word(&mut self) -> Result<u32, Box<dyn Error>> {
        let word = self.reader.read_word(self.offset)?;
        self.offset += 4;
        Ok(word)
    }

    pub fn read_xword(&mut self) -> Result<u64, Box<dyn Error>> {
        let xword = self.reader.read_xword(self.offset)?;
        self.offset += 8;
        Ok(xword)
    }

    pub fn read_int(&mut self) -> Result<i32, Box<dyn Error>> {
        let word = self.reader.read_word(self.offset)?;
        self.offset += 4;
        Ok(word as i32)
    }

    /// Corresponds to the kernel's user_long_t which, I think,
    /// is 64 or 32 bits.
    pub fn read_ulong(&mut self) -> Result<u64, Box<dyn Error>> {
        if self.reader.sixty_four_bit {
            let word = self.reader.read_xword(self.offset)?;
            self.offset += 8;
            return Ok(word);
        } else {
            let word = self.reader.read_word(self.offset)?;
            self.offset += 4;
            return Ok(word as u64);
        }
    }

    pub fn read_addr(&mut self) -> Result<u64, Box<dyn Error>> {
        if self.reader.sixty_four_bit {
            let word = self.reader.read_xword(self.offset)?;
            self.offset += 8;
            return Ok(word);
        } else {
            let word = self.reader.read_word(self.offset)?;
            self.offset += 4;
            return Ok(word as u64);
        }
    }

    pub fn read_offset(&mut self) -> Result<u64, Box<dyn Error>> {
        if self.reader.sixty_four_bit {
            let word = self.reader.read_xword(self.offset)?;
            self.offset += 8;
            return Ok(word);
        } else {
            let word = self.reader.read_word(self.offset)?;
            self.offset += 4;
            return Ok(word as u64);
        }
    }

    /// Read a null-terminated ASCII string.
    pub fn read_string(&mut self) -> Result<String, Box<dyn Error>> {
        let mut s = String::new();
        loop {
            // Kernel documents these as ASCII though I'm not sure I believe that.
            let byte = self.read_byte()?;
            if byte == 0 {
                break;
            }
            s.push(byte as char);
        }
        Ok(s)
    }
}
