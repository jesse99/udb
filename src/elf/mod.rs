//! Generic ELF file support. These can be both core files and executable files.
//! (The executable files are useful because they contain the bulk of the debugging
//! data needed to interpret the core files.) This module contains generic ELF suport
//! but not the support for stuff like interpreting debug symbols.
//! Quick ELF reference: https://gist.github.com/x0nu11byt3/bcb35c3de461e5fb66173071a2379779
//!
//! ELF files start with an ELF header which includes:
//! * A magic number to identify the file as an ELF file.
//! * The architecture, e.g. Linux AMD x86-64.
//! * The offset to and number of program headers.
//! * The offset to and number of section headers.
//!
//! Program headers identify segments. Segments are used by the OS to load an exe into
//! memory. A program header has type, vaddr, offset, etc. Common types are:
//! * Load - for a core file these are usually memory mapped files for the exe and DLLs, for an exe these are text (CPU instructions) and data (eg statics)
//! * Note - variety of metadata, e.g. process and signal info.
//! * TLS - thread local storage info.
//!
//! Section headers identify sections. Sections are used for static linking and don't
//! appear in core files. Section headers have name, type, vaddr, offset, size, etc.
//! There are a lot of types including for the symbol table, string table, etc.
pub mod elf_file;
pub mod elf_files;
pub mod header;
pub mod io;
pub mod notes;
pub mod sections;
pub mod segments;

pub use elf_file::*;
pub use elf_files::*;
pub use header::*;
pub use io::*;
pub use notes::*;
pub use sections::*;
pub use segments::*;
