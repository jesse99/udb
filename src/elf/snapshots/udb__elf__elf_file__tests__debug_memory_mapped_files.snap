---
source: src/elf/elf_file.rs
expression: s
---
/home/vorisek/shopping/app-debug
/usr/lib/x86_64-linux-gnu/libc.so.6
/usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
/usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
